src/app/3d/qgs3daxis.cpp

```diff
@@ -0,0 +1,646 @@
+  Date                 : March 2022
+  Copyright            : (C) 2022 by Jean Felder
+  Email                : jean dot felder at oslandia dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgs3daxis.h"
+
+#include <Qt3DCore/QTransform>
+#include <Qt3DExtras/QConeMesh>
+#include <Qt3DExtras/QCylinderMesh>
+#include <Qt3DExtras/QPhongMaterial>
+#include <Qt3DRender/qcameralens.h>
+#include <Qt3DRender/QCameraSelector>
+#include <Qt3DRender/QClearBuffers>
+#include <Qt3DRender/QLayer>
+#include <Qt3DRender/QLayerFilter>
+#include <Qt3DRender/QPointLight>
+#include<ctime>
+
+#include <Qt3DRender/QAttribute>
+
+Qgs3DAxis::Qgs3DAxis( Qt3DExtras::Qt3DWindow *parentWindow, Qt3DCore::QEntity *parent3DScene, QgsCameraController *cameraCtrl, const Qgs3DMapSettings *map )
+  : QObject( parentWindow ), mParentWindow( parentWindow ), mParentCamera( cameraCtrl->camera() ),
+    mCylinderLength( 40.0f ), mAxisViewportSize( 4.0 * mCylinderLength ),
+    mAxisViewportVertPos( AxisViewportPosition::END ), mAxisViewportHorizPos( AxisViewportPosition::END ),
+    mFontSize( 10 ), mMode( Mode::SRS ), mAxisRoot( nullptr ), mTextTransform_X( nullptr ), mTextTransform_Y( nullptr ),
+    mTextTransform_Z( nullptr ), mCrs( map->crs() )
+{
+  mAxisViewport = constructAxisViewport( parent3DScene );
+  mAxisViewport->setParent( mParentWindow->activeFrameGraph() );
+
+  mTwoDLabelViewport = constructLabelViewport( parent3DScene, QRectF( 0.0f, 0.0f, 1.0f, 1.0f ) );
+  mTwoDLabelViewport->setParent( mParentWindow->activeFrameGraph() );
+
+  connect( cameraCtrl, &QgsCameraController::cameraChanged, this, &Qgs3DAxis::updateCamera );
+  connect( mParentWindow, &Qt3DExtras::Qt3DWindow::widthChanged, this, &Qgs3DAxis::updateAxisViewportSize );
+  connect( mParentWindow, &Qt3DExtras::Qt3DWindow::heightChanged, this, &Qgs3DAxis::updateAxisViewportSize );
+
+  updateAxisViewportSize();
+
+  createAxisScene();
+}
+
+Qt3DRender::QViewport *Qgs3DAxis::constructAxisViewport( Qt3DCore::QEntity *parent3DScene )
+{
+  auto axisViewport = new Qt3DRender::QViewport;
+  // size will be set later
+
+  mAxisSceneEntity = new Qt3DCore::QEntity;
+  mAxisSceneEntity->setParent( parent3DScene );
+
+  mAxisCamera = new Qt3DRender::QCamera;
+  mAxisCamera->setParent( mAxisSceneEntity );
+  mAxisCamera->setProjectionType( mParentCamera->projectionType() );
+  // always ortho for now as it seem more "accurate"
```

Comment:
this looks wrong to me -- I believe it's what's leading to the issue @uclaros pointed out with the axis misalignment

Reply:
see comment https://github.com/qgis/QGIS/pull/48109#issuecomment-1091443114. Thx